ocaml/xcp-rrdd/bin/rrdp-dcmi/rrdp_dcmi.ml

```
@@ -27,31 +27,49 @@ let ipmitool_bin = "/usr/bin/ipmitool"
 
 let ipmitool args =
   (* we connect to the local /dev/ipmi0 if available to read measurements from local BMC *)
-  ipmitool_bin :: "-I" :: "open" :: args |> String.concat " "
+  ipmitool_bin :: args |> String.concat " "
+
+type discovery_error = Devices_missing
+
+let discovery_error_to_string = function
+  | Devices_missing ->
+      "IPMI devices are missing"
 
 let discover () =
+  let read_out_line line =
+    (* this code runs once on startup, logging all the output here will be useful for debugging *)
+    D.debug "DCMI discover: %s" line ;
+    let line = String.trim line in
+    if String.equal line "Power management available" then
+      Some ()
+    else
+      None
+  in
+  let read_err_line line =
+    (* this code runs once on startup, logging all the output here will be useful for debugging *)
+    D.debug "DCMI discover: %s" line ;
+    let line = String.trim line in
+    if String.starts_with ~prefix:"Could not open device at" line then
+      Some Devices_missing
+    else
+      None
+  in
   Utils.exec_cmd
     (module Process.D)
     ~cmdstring:(ipmitool ["dcmi"; "discover"])
-    ~f:(fun line ->
-      (* this code runs once on startup, logging all the output here will be useful for debugging *)
-      D.debug "DCMI discover: %s" line ;
-      if String.trim line = "Power management available" then
-        Some ()
-      else
-        None
-    )
+    ~read_out_line ~read_err_line
 
 let get_dcmi_power_reading () =
+  let read_out_line line =
+    (* example line: '     Instantaneous power reading:                    34 Watts' *)
+    try Scanf.sscanf line " Instantaneous power reading : %f Watts" Option.some
+    with Scanf.Scan_failure _ | End_of_file -> None
```

Would it not be safer to catch all exceptions?